### PR TITLE
Default values for widget model attributes

### PR DESCRIPTION
**Pull request by SylvainCorlay**
@jdfreder @parente here is the approach I am proposing for specifying default values for widget attributes on the javascript side.

I started with the base widget classes and the `_Bool` widgets.

TODO:
- [x] figure out what to do with the instance of layout widget.
- [x] check commit regarding buffered state diff.
- [ ] some widgets have the same models and view types, but *different default values*. Is it worth inheriting and specifying a custom model for these cases (range sliders vs sliders)

This is an API change from the perspective of custom widget authors who need custom models.

### CODE REVIEW

**Comment by jdfreder** [comment]
From the sample you have here, I think this is great.

**Comment by SylvainCorlay** [comment]
Test failures seem unrelated.

File: ipywidgets/static/widgets/js/widget_bool.js [hunk 1]
```diff
@@ -10,8 +10,24 @@ define([
...
+        defaults: _.extend(widget.DOMWidgetModel.prototype.defaults, {
+            value: false,
+            description: "",
+            disabled: false,
+            _model_name: "BoolModel"
+        }),
+    });
+
+    var CheckboxModel = BoolModel.extend({
+        defaults: _.extend(BoolModel.prototype.defaults, {
```
Comment (jdfreder) [inline ipywidgets/static/widgets/js/widget_bool.js]: This is the line that is causing the tests to fail.  You need to import underscore.

**Comment by jdfreder** [comment]
> Test failures seem unrelated.

They are related, missing import, see above

File: ipywidgets/static/widgets/js/widget_bool.js (hunk 1 above)
Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/widget_bool.js]: fixed thanks!

**Comment by SylvainCorlay** [comment]
Thanks. I will iterate on this PR to cover the other models.

File: ipywidgets/static/widgets/js/widget_bool.js (hunk 1 above)
Comment (jdfreder) [inline ipywidgets/static/widgets/js/widget_bool.js]: Thanks the new karma tests ;)

**Comment by SylvainCorlay** [comment]
- The color picker widget has a recently added boolean attribute called `short` specifying whether it should take a full widget width and have a text input area or just have a picker button and have a short widget width. We could rename this to "concise".

**Comment by jdfreder** [comment]
> We could rename this to "concise".

Sure

**Comment by parente** [comment]
Ping @lbustelo

**Comment by SylvainCorlay** [comment]
>> We could rename this to "concise".

> Sure

Done

**Comment by SylvainCorlay** [comment]
I have to figure out how to define a default for the instance of layout widget from the js, the default being to instantiate a new one...

**Comment by jdfreder** [comment]
Great work so far!  Looking forward to this one

**Comment by SylvainCorlay** [comment]
Would love to have your opinion on the `buffered_state_diff` commit.

**Comment by SylvainCorlay** [comment]
The issue was that the state from `defaults` was being added to `buffered_state_diff` at the time when the base constructor was called. Then the default values were sent back to the backend.

I changed the code in the following fashion: only buffer state diff after the initial state has been set.

File: ipywidgets/static/widgets/js/widget.js [hunk 2]
```diff
@@ -548,7 +563,11 @@ define(["./utils",
         }
     };
 
-    var DOMWidgetModel = WidgetModel.extend({}, {
+    var DOMWidgetModel = WidgetModel.extend({
+        defaults: _.extend({}, WidgetModel.prototype.defaults, {
+            layout: undefined,
```
Comment (jdfreder) [inline ipywidgets/static/widgets/js/widget.js]: Isn't this supposed to be `null`?

**Comment by jdfreder** [comment]
@SylvainCorlay one small comment

File: ipywidgets/static/widgets/js/widget.js (hunk 2 above)
Comment (SylvainCorlay) [inline ipywidgets/static/widgets/js/widget.js]: Well I don't know. We don't allow `None` for the layout widget. 
In fact, even when building it from JavaScript, we should instantiate a layout widget. Not sure what is the best way to do it. That is why I used `undefined`.

Comment (jdfreder) [inline ipywidgets/static/widgets/js/widget.js]: That's okay for now, note to self: test edge cases of this.
